core/domain/user_services.py

```diff
@@ -2287,3 +2291,291 @@ def get_dashboard_stats(user_id):
+                state_name)
+        else:
+            # If furthest reached checkpoint is behind most recently
+            # reached checkpoint.
+            if (checkpoints_in_current_exploration.index(
+                furthest_reached_checkpoint_in_current_exploration) <= (
+                    checkpoints_in_current_exploration.index(state_name))):
+                (
+                    exploration_user_model
+                    .furthest_reached_checkpoint_exp_version
+                ) = exp_version
+                (
+                    exploration_user_model
+                    .furthest_reached_checkpoint_state_name
+                ) = state_name
+
+    exploration_user_model.most_recently_reached_checkpoint_exp_version = (
+        exp_version)
+    exploration_user_model.most_recently_reached_checkpoint_state_name = (
+        state_name)
+    exploration_user_model.update_timestamps()
+    exploration_user_model.put()
+
+
+def set_user_has_viewed_lesson_info_modal_once(user_id):
+    """Updates the user's settings once he has viewed the lesson info modal.
+
+    Args:
+        user_id: str. The Id of the user.
+    """
+
+    user_settings = get_user_settings(user_id)
+    user_settings.mark_lesson_info_modal_viewed()
+    _save_user_settings(user_settings)
+
+
+def update_learner_checkpoint_progress_on_restart(user_id, exploration_id):
+    """Sets the most recently reached checkpoint on restart exploration event.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+    """
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+    if exploration_user_model is None:
+        # This case normally won't arise since as soon as the exploration
+        # starts most_recently_reached_checkpoint is updated (First state of
+        # an exploration is always a checkpoint). This is just a double check
+        # in case learner hits the restart button while the other content of
+        # the page is loading.
+        exploration_user_model = user_models.ExplorationUserDataModel.create(
+            user_id, exploration_id)
+    exploration_user_model.most_recently_reached_checkpoint_exp_version = None
+    exploration_user_model.most_recently_reached_checkpoint_state_name = None
+    exploration_user_model.update_timestamps()
+    exploration_user_model.put()
+
+
+def sync_learner_checkpoint_progress_with_current_exp_version(
+    user_id,
+    exploration_id):
+    """Synchronizes the most recently reached checkpoint and the furthest
+    reached checkpoint with the latest exploration.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+
+    Returns:
+        ExplorationUserData. The domain object corresponding to the given user
+        and exploration.
+    """
+    exploration_user_model = user_models.ExplorationUserDataModel.get(
```

Comment:
```suggestion
    exp_user_model = user_models.ExplorationUserDataModel.get(
```

it might help with some of the formattings below

Reply:
ditto elsewhere

Reply:
Done!